### PR TITLE
Add webhook to capture payments

### PR DESCRIPTION
**Pull request by Naokimi**
Fixes https://github.com/nebulab/solidus_bolt/issues/33

~~Bolt request verification has been temporarily commented out and will be implemented in https://github.com/nebulab/solidus_bolt/issues/79~~ Fixed in commit 12df15f5b4e4e609ba30463aa8284955cde9009f

Manual testing results:
Full capture:
https://merchant-sandbox.bolt.com/transaction/B3LQ-97ZW-BLP3
image

Partial capture:
https://merchant-sandbox.bolt.com/transaction/W2YD-RLG4-DCP2
image

### CODE REVIEW

**Review by DanielePalombo — CHANGES_REQUESTED** [review]

File: app/webhooks/solidus_bolt/handlers/base.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+module SolidusBolt
+  module Handlers
+    class Base
```
Comment (DanielePalombo) [inline app/webhooks/solidus_bolt/handlers/base.rb:5]: Thank you to have moved the Handlers under webhooks path, to reflect the RailsWay, do you mind adding `Handler` suffix to all handlers?

File: app/services/solidus_bolt/payments/capture_service.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+      def initialize(payment:)
+        @payment = payment
+        @order = payment.order
+        @payment_source = payment.payment_source
+        super
+      end
+
+      def call
+        return unless payment_source.can_capture?(payment)
```
Comment (DanielePalombo) [inline app/services/solidus_bolt/payments/capture_service.rb:16]: I don't think we should check it. 
Considering that Bolt is the source of truth if it says that the payment has been captured, we just have to move the payment to the capture state.

Comment (DanielePalombo) [reply app/services/solidus_bolt/payments/capture_service.rb:16]: Following the above conversion, maybe we just have to check if the payment is already captured to avoid capturing an already captured payment.
It will happen whenever the payment will be captured from the solidus side, and the webhook is enabled.

**Review by DanielePalombo — CHANGES_REQUESTED** [review]

File: app/services/solidus_bolt/payments/base_service.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+module SolidusBolt
+  module Payments
+    class BaseService
```
Comment (DanielePalombo) [inline app/services/solidus_bolt/payments/base_service.rb:5]: Why did you add that class?

**Review by DanielePalombo — CHANGES_REQUESTED** [review]
Probably "request" tests here will make the test suite more strong.

**Comment by DanielePalombo** [comment]
Does it work for you?

I've got two main errors:

1 `ActionView::Template::Error (Missing partial spree/api/payments/source_views/_bolt with {:locale=>[:en], :formats=>[:json], :variants=>[], :handlers=>[:jbuilder]}` when the webhook is received
2 I had to skip verify_bolt_request check

A part these, when the webhook is received the payment is captured as expected.

**Comment by Naokimi** [comment]
tested by calling the controller using Postman and confirmed the code works as expected

**Comment by DanielePalombo** [comment]
Why you didn't test it with Bolt webhooks?

File: app/services/solidus_bolt/payments/capture_service.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+module SolidusBolt
+  module Payments
+    class CaptureService < SolidusBolt::BaseService
```
Comment (piyushswain) [inline app/services/solidus_bolt/payments/capture_service.rb:5]: Would moving this and related class under `SolidusBolt::Webhooks::Payments::CaptureService` make more sense ?

Since this service seems specific to Webhook requests

Comment (DanielePalombo) [reply app/services/solidus_bolt/payments/capture_service.rb:5]: I don't think that should exist `Sebhook services`. A webhook is just a way to receive a notification, and it should just have to call a service.

**Review by piyushswain — APPROVED** [review]

**Review by DanielePalombo — CHANGES_REQUESTED** [review]

File: app/services/solidus_bolt/payments/capture_service.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        super
+      end
+
+      def call
+        return if payment.completed?
+
+        if order_amount == capture_amount
+          payment.complete!
+        else
+          Spree::PaymentCaptureEvent.create(payment: payment, amount: Money.new(capture_amount).to_d)
```
[the remark this answers is not in the document]
Comment (Naokimi) [reply app/services/solidus_bolt/payments/capture_service.rb:21]: the core gist of the method is creating a [capture_event for the amount captured](https://github.com/spree/spree/blob/ab448523bdc112306b67e73bcf16ef132e791df7/core/app/models/spree/payment/processing.rb#L45), which is what I'm doing in the partial capture

the part above (lines 39-43) call the Gateway to make a capture request, which is what we want to avoid since the payment was already captured on Bolt
```
    24: def capture(float_amount, response_code, gateway_options)
    25:   binding.pry
 => 26:   payment = gateway_options[:originator]
```

[the remark this answers is not in the document]
Comment (DanielePalombo) [reply app/services/solidus_bolt/payments/capture_service.rb:21]: Ok, now i see your point. But there is also a method called 'split_uncaptured_amount' that should executed anyway. In that way the payment amount will be updated and a new payment with the remaining amount will be created

File: app/controllers/solidus_bolt/webhooks_controller.rb [hunk 6]
```diff
@@ -20,11 +21,12 @@ def update
     private
 
     def verify_bolt_request
-      hash_header = request.headers['X-Bolt-Hmac-Sha256']
+      hmac_header = request.headers['X-Bolt-Hmac-Sha256']
       signing_secret = SolidusBolt::BoltConfiguration.fetch&.signing_secret || ''
-      bolt_request = hash_header == Base64.encode64(OpenSSL::HMAC.digest("SHA256", params.to_json, signing_secret))
+      payload = params.permit!.to_h[:webhook].to_json
```
Comment (DanielePalombo) [inline app/controllers/solidus_bolt/webhooks_controller.rb:26]: Why you have used the permit! method? Permitted attributes should have used only when the params is provided to an activerecord method that updates the record. I'm worried the executing permit! On this params will make the whole object permitted, even if it shouldn't be permitted under the handler service. WDYT?

Comment (DanielePalombo) [reply app/controllers/solidus_bolt/webhooks_controller.rb:26]: I think that params[:webhook] should be enough

File: app/services/solidus_bolt/payments/capture_service.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+module SolidusBolt
+  module Payments
+    class CaptureService < SolidusBolt::BaseService
```
Comment (DanielePalombo) [inline app/services/solidus_bolt/payments/capture_service.rb:5]: CaptureService doesn't capture the payment. This would make a lot of confusion. Can you rename it?

**Review by DanielePalombo — CHANGES_REQUESTED** [review]

File: app/services/solidus_bolt/payments/capture_service.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (Naokimi) [reply app/services/solidus_bolt/payments/capture_service.rb:21]: Interesting. I based my code on the [snippet from Solidus](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/payment/processing.rb#L46-L67) instead of the one from Spree that you are sharing, and Solidus doesn't have that `split_uncaptured_amount` method

File: app/services/solidus_bolt/payments/capture_service.rb (hunk 7 above)
Comment (Naokimi) [reply app/services/solidus_bolt/payments/capture_service.rb:5]: `CompleteOrPartialCaptureService` is a bit of a mouthful tho 🤔

**Review by DanielePalombo — APPROVED** [review]
LGTM
